src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1755,206 +1906,299 @@ public static decimal Sum<TSource>(this IEnumerable<TSource> source, Func<TSourc
 
         public static int Min(this IEnumerable<int> source)
         {
-            if (source == null) throw Error.ArgumentNull("source");
-            int value = 0;
-            bool hasValue = false;
-            foreach (int x in source)
+            Error.CheckSourceNotNull(source);
+            int value;
+            using (IEnumerator<int> e = source.GetEnumerator())
             {
-                if (hasValue)
+                if (!e.MoveNext()) throw Error.NoElements();
+                value = e.Current;
+                while (e.MoveNext())
                 {
+                    int x = e.Current;
                     if (x < value) value = x;
                 }
-                else
-                {
-                    value = x;
-                    hasValue = true;
-                }
             }
-            if (hasValue) return value;
-            throw Error.NoElements();
+            return value;
```

Comment:
> if you are coding side-effects into your Linq use you are going to have a bad time, but that doesn't mean people won't have done so

Unfortunately people definitely do this.

> I haven't added short-circuits where there isn't a test in the original

We have very low coverage of LINQ in our tests right now. Please don't infer that a lack of a test for a particular behavior means that behavior can be changed (though it'd be nice if we lived in such a world :smile:).

> Perhaps whether or not to short-circuit needs to be discussed separately.

Yes :smile:   I think it'd be best to separate out anything that could result in a functional change in behavior.  Otherwise, such things could very easily derail other suggested improvements.